### ISSUE TITLE
Large top_k (> SAMPLE_MAX_TOP_K) degenerates: CUB sampling path emits a comma loop

### ISSUE THREAD

**Issue by kekzl** [kekzl/imp#1142]
## Symptom

`top_k = 2000` on Qwen3-8B-Q8_0 produces a repetition loop for the entire token
budget:

```
Okay,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,</think></think></think>,</think></think>,...
```

Reproduce (server on `Qwen3-8B-Q8_0.gguf`):

```bash
curl -s localhost:8080/v1/chat/completions -H 'Content-Type: application/json' -d '{
  "model":"Qwen3-8B-Q8_0.gguf",
  "messages":[{"role":"user","content":"Name three seas."}],
  "max_tokens":120, "top_k":2000, "temperature":0.9, "seed":42}'
```

## Scope

- `top_k = 40` — i.e. **below** `SAMPLE_MAX_TOP_K = 128`, the multiblock path —
  is perfectly coherent on the same build, same seed, same prompt.
- Above 128 the sampler switches to `sample_topk_topp_cub()`
  (`src/compute/sampling_topk_topp.cu`), which is where the defect must be.
- Not a regression: verified against a `main`-built image, which reproduces the
  output byte-for-byte. Found while probing that path for unrelated memory work
  (AUDIT B76).

## Why nothing caught it

`tools/analysis/degen_suite.py` passes 41/41 on the same server — nothing in the
battery samples above `SAMPLE_MAX_TOP_K`, so the CUB path has no coverage at
all. Whatever the fix, it wants a battery case with a large `top_k`.

## Suspects, in the order worth checking

1. The softmax stats (`d_max_sum`) or the top-p prefix scan over the CUB-sorted
   output — a wrong sum makes the cutoff select from the tail.
2. The `DeviceTopK::MaxPairs` result ordering: it is requested with
   `output_ordering::unsorted`, and the top-p step downstream assumes a
   descending order that a later `SortPairsDescending` is supposed to establish.
3. `top_k` larger than the vocabulary slice actually sorted.

**Comment by kekzl** [comment]
Investigated. **The title's premise was wrong** — this is not the CUB path
mis-sampling. Updating with what is now established, so the next pass starts
from facts.

## The boundary is exactly 128/129

Server on `Qwen3-8B-Q8_0.gguf`, same prompt, same seed, `temperature 0.9`:

| top_k | output |
|---|---|
| 40 | `Okay, the user is asking for three seas. Let me think about the most commonly known ones f…` |
| 100 | coherent |
| 128 | coherent |
| **129** | `Okay,,,,,,,,,,,,,,,,,,,</think></think>…` |
| 500 | `Okay</think>\n\n</think></think>,…` |

128 is `SAMPLE_MAX_TOP_K`, i.e. the multiblock ↔ CUB path split.

## Ruled out

- **CUB temp-storage sizing.** The scratch queries CUB's requirement once with
  `k = vocab_size` and calls with the request's `k`. Measured at V=151936:
  `DeviceTopK` is flat at 28 671 B for k ∈ {128, 2000, 20000, V}, and the
  reserved temp is the max over all three queries. Amply sized.
- **The CUDA-graph decode path.** Reproduces identically with
  `--set runtime.cuda_graphs=never`.
- **The think budget.** Reproduces identically with `think_budget: 0`
  (`</think>` spam in the output made this the obvious suspect; it is not it).
- **The CUB sampler itself, on synthetic logits.** At the real vocabulary
  (V=151936) with a clear top-3, `top_k` 128 / 129 / 2000 all put 90–95 % of 100
  draws on the winners. The kernel samples correctly when driven directly.
- **Seed correlation** — this one was a real bug and is fixed in #1143, but it
  is *not* this one: the degeneration survives that fix.

## Where that leaves it

Something on the **engine-side path taken for `top_k > SAMPLE_MAX_TOP_K`**
differs from the ≤128 path in a way the sampler kernel itself does not explain.
Two candidates worth checking first, both about the declined-async fallback:

1. `sample_topk_topp_async()` returns `false` for `top_k > MAX_TOP_K`, and both
   call sites in `exec/executor.cu` (lines ~356 and ~570) discard it with
   `(void)ok;  // eligibility pre-checked above`. The batched caller in
   `engine_scheduler.cpp` does re-sample declined rows synchronously, so that
   one looks safe — but the claim "cannot decline here" should be turned into an
   assertion rather than a comment, and the single-sequence paths checked the
   same way.
2. The filters applied before sampling (`apply_row_filters_` / penalties /
   min_p) run in a different order on the two paths; a filter applied twice or
   not at all would produce exactly this.

Repro is one curl (above), so both are cheap to check.

**Comment by kekzl** [comment]
Correction to my own candidate 2 above: **the filters are not applied
differently.** Both `sample_single_from_logits` (sync/CUB) and
`sample_single_from_logits_async` call the same `apply_row_filters_()`, in the
same order — penalties, DRY, banned tokens, logit bias, force-token, constraint
mask, min_p, typical_p. Ruled out.

Better-grounded candidate, and it fits the 128/129 split precisely:

**The two paths treat filtered-out logits differently.** `apply_row_filters_`
writes `-INF`/`-FLT_MAX` into masked positions. The multiblock path selects its
top-k with an explicit comparison that carries an `s_idx[i] >= 0` guard through
the finalize kernel, so masked entries are structurally excluded. The CUB path
instead builds a full-vocabulary probability array
(`softmax_to_pairs_device_kernel`) and hands it to
`cub::DeviceTopK::MaxPairs` + `SortPairsDescending` — and a radix sort over
float keys orders **NaN above +INF**. Any position that comes out of the
softmax as NaN (e.g. `x - max` with both `-inf`, which happens as soon as the
running max over a masked region is itself `-inf`) is therefore selected *first*
by the top-k, and the sampler then draws from a candidate list whose head is
garbage — deterministically, every step, which is what the repeated token looks
like.

Cheap way to confirm before writing any fix: dump `sc.d_keys_out[0..8]` and
`sc.d_vals_out[0..8]` for one decode step at `top_k=129` on the real model. If
the head of the sorted list is NaN, that is the bug and the fix is to sanitise
the probability array (or to select with a comparator that rejects non-finite
keys) rather than anything about `top_k`.

**Comment by kekzl** [comment]
Fixed in #1144 (merged `acc40ffc`). Root cause was `cub::DeviceTopK::MaxPairs` serving stale candidates — it fills its output on the first call, writes nothing on the second while returning `cudaSuccess`, and fails permanently with `invalid device ordinal` from the fourth; nobody checked the return code. Replaced by a single full-vocabulary descending sort, which the scratch was already sized for. `degen_suite.py` gained top_k 129 and 2000 cases — the path had no coverage at all, which is why it shipped.
